### ISSUE TITLE
Coregionalization with heteroscedastic noise

### ISSUE THREAD

**Issue by doccosmos** [SheffieldML/GPy#300]
What is the simplest way to include heteroscedastic noise models with coregionalization? Is it just a case of setting the variance of the various kernel components (one per output) to the appropriate noise value?

**Comment by lawrennd** [comment]
Hi Matthew,

Just to clarify you mean heteroscedastic across dimensions, not data points
right? Tagging in @ric70x7 who's done the most work with heteroscedastic &
coregionalisation in GPy. Ricardo, did you create a notebook for
coregionalisation with different output noises?

Neil

**Comment by doccosmos** [comment]
I have a number of different outputs and each output is associated with a separate heteroscedastic noise model - is this dimensions or data points?

**Comment by lawrennd** [comment]
dimensions of the data is the number of outputs (denoted p in statistics).
Data points is the number of data observations (denoted n). Although it can
be that you have different numbers of observations for each output.

Heteroscedasticity can be modelled acros the different outputs and/or
across different data observations.

Neil

**Comment by mzwiessele** [comment]
There is a notebook in the GPy notebook repository over at Github. It is in the GPy section, if I am not completely mistaken. I am not sure how well it describes heteroscedasticity across dimensions, though. Have a look at the notebook and feel free to ask questions here :)

**Comment by ric70x7** [comment]
Hi Matthew,

A coregionalized model with heteroscedastic noise, that sounds interesting.
It shouldn't be difficult to implement. I wonder how it would perform in
practice.

The first step I recommend you is to go through the tutorials of
heteroscedastic regression and coregionalized regression. Then you can
build your own model.

The core ideas of how this models are implented in GPy are the following:

Heteroscedastic noise: Each of the N observations y_i are associated to a
different noise term sigma_w where w <= N.

y_i = f(x_i) + epsilon[w]_i

When you define your likelihood, you have to define the number of noise
terms in the model. You do this with a dictionary (Y_metadata) that
contains a vector of n indices that associates each y_i to one sigma_w.
Something like this:

Y_metadata = {'output_index':np.arange(Ny)[:,None]}

Then the likelihood can be defined as:

likelihood = likelihoods.HeteroscedasticGaussian(Y_metadata)

If you want the noise terms to be non-Gaussian, or some to be Gaussian and
others not... well that is indeed possible, but the implementation needs
more thinking.

Coregionalized regression: Here you have D outputs, so you can index your
observations as y_di. You don't need to have the same number of
observations for each output. You have something like this:

y_di = f_d(x_i) + \epsilon_di

This model uses a different noise term across outputs, but is homoscedastic
within ouputs. Notice that all y_di have the same input space x_i,
regardless the output d. The way this is handled in GPy is with a
coregionalized kernel, that expands the input dimension (x_d) with a set of
indices that tells to which output d the input point x_n will be used:

x_dn = (x_n | d)

Don't use any kernel directly on these inputs, you have to use a
coregionalized kernel:

K=GPy.kern.RBF(1)B = GPy.kern.Coregionalize(input_dim=1,output_dim=2)
coreg_kernel = K.prod(B,name='B.K')

In summary: Create a coregionalized kernel.

Create a likelihood with the Y_metadata defining your different noise terms.

Append to the right of X a column of output indices.

Put everything together in a regression model, and blend...

Don't try to use the GPy.models.GPCoregionalizedRegression, it has some
pre-defined parameters that won't work with your assumptions.

You can use the GPHeteroscedasticeRegression, but passing as parameters of
model a Coregionalized kernel and an already expanded input dimension set X.

Good luck,

Ricardo

**Comment by doccosmos** [comment]
In the coregionalized tutorial, the regression class takes a list of inputs (X_list), one per output, but you seem to be saying that GPHeteroscedasticRegression needs to take in a single set of inputs which is the same for all outputs. Do I need to construct a superset of all the different inputs for heteroscedastic regression and tell the kernel which point goes with each output?

**Comment by ric70x7** [comment]
Yes, the coregionalization model does the indexing and expansion of the
input set for you. But that is not the way the upper class of regression
model works.
You need:
-  an array Y (N x 1), where N = N_1 + N_2 + ... + N_D; N_d is the number
  of observations of output d, and D is the total number of outputs.
- an array X (N x (q+1) ), where q is the actual input dimension, and the
  extra 1 is the index that has the information about the output.

You can use this function to build the arrays:

GPy.util.multioutput.build_XY(input_list,output_list=None,index=None)

2016-02-09 16:19 GMT-08:00 Matthew Graham notifications@github.com:

**Comment by doccosmos** [comment]
I'm getting the following error when I try and follow this. Can you also comment on why the GPCoregionalizedRegression is not appropriate when the class description says that it is a "Gaussian Process model for heteroscedastic multioutput regression".

def getTestData():
  npts = np.random.randint(300)
  dt = np.abs(np.random.normal(25., 10., npts - 1))
  t = np.zeros(npts)
  t[1:] = np.cumsum(dt)
  obs = np.random.normal(0., 1., npts)
  errseed = np.random.normal(0., 0.05, npts)
  err = np.abs(np.random.normal(0., np.abs(np.random.normal(0., 0.05, npts))))
  return np.array([t, obs, err])

### Get data

nExp = 2
input_list = []
output_list = []
variance = np.array([])

### For each expt

for i in range(len(expt)):
  t, obs, err = getTestData()
  input_list.append(t[:, None]) 
  output_list.append(obs[:, None])
  variance = np.concatenate((variance, err *\* 2.))

### Construct correct input and output data constructs

X, Y, idx = GPy.util.multioutput.build_XY(input_list, output_list)
Y_metadata = {'output_index': idx}

### Define coregionalized kernel

K = GPy.kern.OU(1)
B = GPy.kern.Coregionalize(input_dim = 1, output_dim = nExp)
coreg_kernel = K.prod(B, name = "B.K")

### Fit GP OU model

m = GPy.models.GPHeteroscedasticRegression(X, Y, kernel = coreg_kernel, Y_metadata = Y_metadata)

---

LinAlgError                               Traceback (most recent call last)
<ipython-input-248-20d8880a49fb> in <module>()
     53 
     54 # Fit GP OU model
---> 55 m = GPy.models.GPHeteroscedasticRegression(X, Y, kernel = coreg_kernel, Y_metadata = Y_metadata)
     56 m['.*het_Gauss.variance'] = variance[:, None] # Set the noise parameters to the error in Y
     57 m.het_Gauss.variance.fix() # We can fix the noise term, since we already know it

/Users/mjg/anaconda/lib/python2.7/site-packages/GPy/core/parameterization/parameterized.pyc in **call**(self, _args, *_kw)
     25         self._highest_parent_._connect_fixes()
     26         logger.debug("calling parameters changed")
---> 27         self.parameters_changed()
     28         return self
     29 

/Users/mjg/anaconda/lib/python2.7/site-packages/GPy/core/gp.pyc in parameters_changed(self)
    186             this method yourself, there may be unexpected consequences.
    187         """
--> 188         self.posterior, self._log_marginal_likelihood, self.grad_dict = self.inference_method.inference(self.kern, self.X, self.likelihood, self.Y_normalized, self.mean_function, self.Y_metadata)
    189         self.likelihood.update_gradients(self.grad_dict['dL_dthetaL'])
    190         self.kern.update_gradients_full(self.grad_dict['dL_dK'], self.X)

/Users/mjg/anaconda/lib/python2.7/site-packages/GPy/inference/latent_function_inference/exact_gaussian_inference.pyc in inference(self, kern, X, likelihood, Y, mean_function, Y_metadata)
     54         Ky = K.copy()
     55         diag.add(Ky, likelihood.gaussian_variance(Y_metadata)+1e-8)
---> 56         Wi, LW, LWi, W_logdet = pdinv(Ky)
     57 
     58         alpha, _ = dpotrs(LW, YYT_factor, lower=1)

/Users/mjg/anaconda/lib/python2.7/site-packages/GPy/util/linalg.pyc in pdinv(A, _args)
    206 
    207     """
--> 208     L = jitchol(A, *args)
    209     logdet = 2._np.sum(np.log(np.diag(L)))
    210     Li = dtrtri(L)

/Users/mjg/anaconda/lib/python2.7/site-packages/GPy/util/linalg.pyc in jitchol(A, maxtries)
     63         diagA = np.diag(A)
     64         if np.any(diagA <= 0.):
---> 65             raise linalg.LinAlgError("not pd: non-positive diagonal elements")
     66         jitter = diagA.mean() \* 1e-6
     67         num_tries = 1

LinAlgError: not pd: non-positive diagonal elements

**Comment by ric70x7** [comment]
Ok, I think I got confused about what you need. Sorry, it was quite early
here, when I read your email.

Lets go from the beginning. The intention of using the term heteroscedastic
in the descripton of the coregionalized_regression class is to make clear
that each output has a different noise term. So the model is
heteroscedastic across outputs, and there is a noise term for each output.
After looking at your code, I think that this is what you want. If that is
the case, ignore everything I've said and use the coregionalized regression
model.

But, if you want to build a coregionalization model, where you have more
noise terms than outputs, then you have to follow my previous
instructions... and have some good data quality, so that the model is able
to learn what is being asked.

Now, the error you got is because the covariance matrix is not positive
definite. Try playing with the lenghtscale (use a larger one) or add
something to the diagonal.

Cheers,

Ricardo

2016-02-09 18:16 GMT-08:00 Matthew Graham notifications@github.com:

**Comment by doccosmos** [comment]
OK, so what is the equivalent call to:

m['.*het_Gauss.variance'] = variance[:, None]

for the GPCoregionalizedRegression?

**Comment by ric70x7** [comment]
The noise parameters are named as :

mixed_noise.Gaussian_noise_j.variance, with j = 0, 1, ...

m['.*noise.variance'] should work.

2016-02-10 0:31 GMT-08:00 Matthew Graham notifications@github.com:

**Comment by doccosmos** [comment]
Unfortunately that's giving an attribute error:

K = GPy.kern.OU(1)
icm = GPy.util.multioutput.ICM(input_dim = 1, num_outputs = nSurvey, kernel = K)
m = GPy.models.GPCoregionalizedRegression(input_list, output_list, kernel = icm)

---> 59 m['.*noise.variance'] = variance[:, None]
     60 m.OU.variance.fix()
     61 m.optimize()

/Users/mjg/anaconda/lib/python2.7/site-packages/GPy/core/parameterization/parameterized.pyc in **setitem**(self, name, value, paramlist)
    303             self.trigger_update()
    304         else:
--> 305             try: param = self.**getitem**(name, paramlist)
    306             except: raise
    307             param[:] = value

/Users/mjg/anaconda/lib/python2.7/site-packages/GPy/core/parameterization/parameterized.pyc in **getitem**(self, name, paramlist)
    284             if paramlist is None:
    285                 paramlist = self.grep_param_names(name)
--> 286             if len(paramlist) < 1: raise AttributeError(name)
    287             if len(paramlist) == 1:
    288                 if isinstance(paramlist[-1], Parameterized):

AttributeError: .*noise.variance

**Comment by mzwiessele** [comment]
Please have a look at the notebook basic model handling in the Notebook repository from SheffieldML there it will tell you how to set and get parameters. Maybe you do not want to use regular expressions as it is an old feature. You can set and get parameters directly, they behave like numpy arrays :) So I think, not tested, you can go to m.likelihood.variances[:] = <var> and set them all like setting a numpy array. It might be, that the variances are called variance as well, so just have a play with it. It is actually quite intuative, once you get the hang of it :)

**Comment by doccosmos** [comment]
It does not look like you can set the variances for GPCoregionalizedRegression to reflect heteroscedasticity:

m.mixed_noise.Gaussian_noise_0.variance = variance0[:, None]
m.mixed_noise.Gaussian_noise_1.variance = variance1[:, None]

gives

ValueError: could not broadcast input array from shape (165,1) into shape (1)

The same holds for m.likelihood.Gaussian_noise_*.variance

**Comment by ric70x7** [comment]
Please have a look at the coregionalized regression tutorial.
http://nbviewer.jupyter.org/github/SheffieldML/notebook/blob/master/GPy/coregionalized_regression_tutorial.ipynb
You can use a different noise term (but just one) per output.

If what you need is heteroscedasticity within the ouputs, the
coregionalized regression model won't be useful to you. Such a model does
not have a friendly wrapper already coded in GPy. The tools are there to
define this model, but it is not straightforward. You would need to follow
the instructions I sent before and also have a look at this tutorial:
http://nbviewer.jupyter.org/github/SheffieldML/notebook/blob/master/GPy/heteroscedastic_regression.ipynb

Also, please be aware that at the moment GPy does not have a routine to
learn a functional form of the noise. You can only learn noise parameters
at the training locations. Noise prediction at new points is not
implemented.

2016-02-16 13:43 GMT-08:00 Matthew Graham notifications@github.com:

**Comment by doccosmos** [comment]
OK, the confusion was because you had implied above from my sample code that the CoregionalizedRegression model was the one I wanted to be using. To be specific, the current coregionalized regression is only heteroscedastic in assuming a different **single** variance per output (which should possibly be stated in some high level description of it).

**Comment by ric70x7** [comment]
Yes, that's it.

**Comment by doccosmos** [comment]
OK, I've reimplemented as the code above but no matter what the sizes of the random test data inputs, 
I always get the error that the covariance matrix is not positive definite. Digging into the code, the problem lines are (in ExactGaussianInference.inference)

K = kern.K(X)

which results in a matrix with some machine error negative values (which should be zero). np.diag(K) then is not all positive.

For interest's sake, I tried:

t, obs, err = getTestData(10)
X, Y, idx = GPy.util.multioutput.build_XY([t[:, None]], [obs[:, None]])
K = GPy.kern.OU(1)
B = GPy.kern.Coregionalize(input_dim = 1, output_dim = 1)
coreg_kernel = K.prod(B, name = "B.K")

K0 = K.K(X)
K1 = coreg_kernel.K(X)
print np.diag(K0), np.diag(K1)

[ 1.  1.  1.  1.  1.  1.  1.  1.  1.  1.] 
[  0.00000000e+000   2.15827320e-314   0.00000000e+000   0.00000000e+000
   2.17763814e-314   0.00000000e+000   0.00000000e+000  -9.18719151e-172
   0.00000000e+000   0.00000000e+000]

This does not seem right - surely the result should be the same here in both cases?

**Comment by ric70x7** [comment]
Indeed, that doesn't seem right.
There was a time when those lines of code would do what you were expecting,
but not any more. I wonder how long this bug has been hidden in my
tutorial... it has been for a long time, though.

Anyway, the problem is this line:

B = GPy.kern.Coregionalize(input_dim = 1, output_dim = 1)

We have to add the parameter: active_dims = [1], so you write

B = GPy.kern.Coregionalize(input_dim = 1, output_dim = 1, active_dims = [1])

Here you are telling GPy that whenever you pass an input X to the kernel B,
it will only use its second dimension (i.e., column number one). By
default, whenever you define a kernel, the active dimensions used start
form column 0 up to the number of input dimensions (input_dim).

So that will fix the problem. Thanks for your patience :-)

2016-02-17 16:20 GMT-08:00 Matthew Graham notifications@github.com:
